Basic card component configuration

AndresMurguido001 opened the pull request:
Card, Card Group, Card Header, Card Footer, Card Image

added Card transitions, refactored

Integrated box/flex components for stories

AndresMurguido001 committed:
Basic card component configuration

Card, Card Group, Card Header, Card Footer, Card Image

added Card transitions, refactored

Integrated box/flex components for stories

Separated Card styles / types, refactored Card to use Box

moved / renamed card types file

integrated typeography component for CardHeader, title, subtitle props

added readme and Card types

aromanarguello commented on src/Card/Card.component.tsx:
```diff
@@ -0,0 +1,115 @@
+import CardImage, { ImageProps } from "./CardImage.component"
+import CardFlipper, { CardFlipperProps } from "./CardFlipper.component"
+
+interface CardComposition {
```
This should be `ICardComposition` in general best practices for interface naming is prefixed by the letter `I'

aromanarguello commented on src/Card/Card.component.tsx:
```diff
@@ -0,0 +1,115 @@
+  Flipper: React.FC<CardFlipperProps>
+}
+
+export interface CardProps extends React.HTMLAttributes<HTMLDivElement> {
```
Same here, this should `ICardProps`

aromanarguello commented on src/Card/Card.component.tsx:
```diff
@@ -0,0 +1,115 @@
+import * as React from "react"
+import styled, { css } from "styled-components"
```
Styles should live on their own file. Basically there should be a `Card.styles.ts` with all the `styled-components` and those should be imported into `Card.component.tsx`

aromanarguello commented on src/Card/Card.component.tsx:
```diff
@@ -0,0 +1,115 @@
+  hoverable?: boolean
+}
+
+const CardContainer = styled.div<CardProps>`
```
For example, this should go on the styles file. 

You might want to look into using the extending the box into styled instead of using a `div`.  It would be something like `styled(Box)`. `Box` already has a lot of methods from `styled-system` thus you wouldn't need to add that into the `styled-component`. For instance, because it extends `Box` you can jus call `<Card flex="content" flexDirection="column" padding="auto margin"/>` just as an example. If you want we can go more in depth with this point on slack. Feel free to reach out :)

aromanarguello commented on src/Card/Card.component.tsx:
```diff
@@ -0,0 +1,115 @@
+}: CardProps): JSX.Element => {
+  return (
+    <CardContainer
+      className={`__card__ ${className || ""}`}
```
Not a huge fan of dunders `__` I think we should keep it camel case or with a regular dash. Up for discussion?

aromanarguello commented on src/Card/CardBody.component.tsx:
```diff
@@ -0,0 +1,30 @@
+  bordered?: boolean
+}
+
+const SCardBody = styled.div<CardBodyProps>`
```
I would call this `StyledCardBody` if anything. If not just `Container`

aromanarguello commented on src/Card/CardFlipper.component.tsx:
```diff
@@ -0,0 +1,87 @@
+  enabled?: boolean
+}
+
+const SSide = styled.div<CardFlipperProps>`
```
Same here `StyledSide` maybe? consider using box as well

aromanarguello commented on src/Card/CardHeader.component.tsx:
```diff
@@ -0,0 +1,26 @@
+    `};
+`
+const CardHeader: React.SFC<CardHeaderProps> = ({ children, ...rest }) => (
+  <SCardHeader {...rest}>{children}</SCardHeader>
```
Same lol `StyledCardHeader` maybe?

aromanarguello commented on src/Card/CardImage.component.tsx:
```diff
@@ -0,0 +1,91 @@
+   * Size of the image
+   */
+  size?: ImageSize
+  src: string
```
Did you leave the `jsDoc` out on purpose?

aromanarguello commented on src/Card/CardImage.component.tsx:
```diff
@@ -0,0 +1,91 @@
+  space?: ImageSpaceDirection
+}
+
+const getSize = (size: ImageSize): string => {
```
Would highly encourage looking into the object look up method. I prefer it over switch statements. They have less computational overhead as well. 

```jsx
const retrunSomething =(arg) => {
  const sizes = {
    small: '150px'
  }
  return sizes[arg]
}
```

martindidiego commented on src/Card/Card.component.tsx:
```diff
@@ -0,0 +1,115 @@
+  background: #fff;
+  border-radius: 5px;
+  overflow: hidden;
+  box-shadow: 2px 2px 5px rgba(0, 0, 0, 0.1);
```
Styles look good! Once we add a `theme` object, we definitely want to make sure we go back to each component using literal style values like this one and use `style-system` props instead. This will help the components be more consistent and dynamically update when/if the theme changes!

martindidiego commented on src/Card/Card.component.tsx:
```diff
@@ -0,0 +1,115 @@
+  image,
+  children,
+  ...rest
+}: CardProps): JSX.Element => {
```
You may not need the explicit return value `JSX.Element` here -- `React.FC` already defines that, no?

martindidiego commented on src/Card/CardFlipper.component.tsx:
```diff
@@ -0,0 +1,87 @@
+import styled, { css } from "styled-components"
+import { CardProps } from "./Card.component"
+
+export interface CardFlipperProps extends React.HTMLAttributes<HTMLDivElement> {
```
Nice job on making sure we support all valid HTML attributes for this component!

martindidiego commented on src/Card/CardFlipper.component.tsx:
```diff
@@ -0,0 +1,87 @@
+   * @default false
+   * */
+  vertical?: boolean
+  children: React.ReactElement<CardProps>[]
```
Any particular reason you needed to add this prop explicitly? `React.FC` should automatically support this prop, right?

sherwino replied:
No longer submitting CardFlipper with the native card component.

martindidiego reviewed (commented):
Overall looks good man, nice job. Agreed with some of @aromanarguello's comments and left a couple extra questions.

AndresMurguido001 commented:
Appreciate the reviews guys. I went ahead and refactored according to your comments. I separated the styles into their own file, but it created a circular import error according to eslint, which i why i ended up separating the types into their own file as well. I was thinking of having CardHeader directory / CardFooter directory / Card whatever directory etc... Lmk what you guys think

aromanarguello commented on src/Card/Card.types.tsx:
```diff
@@ -0,0 +1,106 @@
+// Card.component
```
I do like the types living in their own file. Files should be called either `index.d.ts` or `Cards.d.ts`. Thoughts ? @martindidiego @fhern077

martindidiego replied:
Don’t mind either name as long as we’re consistent. `index.d.ts` would be following convention, though, so I prefer that one if I had to choose.

fhern077 replied:
I think putting in ```Cards.d.ts``` would be best and then barrel all types to the ```index.d.ts``` that way it makes it easier to find relevant types + we can comply with convention.

aromanarguello replied:
Yup this should be called `Cards.d.ts`. It will interpret it as the declaration file instead of `Cards.types.tsx` @AndresMurguido001

sherwino replied:
Styled system changed how types are declared.

aromanarguello reviewed (commented):
Declaration file name still up for discussion

AndresMurguido001 committed:
Resolved conflicts

AndresMurguido001 committed:
refactored using styled-system and simplified card

AndresMurguido001 committed:
simplified basic usage example

AndresMurguido001 committed:
modified card stories

AndresMurguido001 committed:
removed 'any' types

AndresMurguido001 commented:
So after seeing the other components added i got a better idea of what the card should be. I think i made it a lot more customizable using styled-system. I also added a Box / Flex components that can be used to replace rebass/grid components. If this looks good to you guys, we can remove that dependency. Also added a type to the returnStyled function bc me and sherwin thought it was interfering with the build process.

sherwino reviewed (approved):
Everything looks great man, 100% test coverage and even without a design these cards look good. 
I saw that you refactored everything to address the comments too!

<img width="995" alt="Screenshot 2019-08-06 11 07 49" src="https://user-images.githubusercontent.com/2348227/62551850-b1ac8c00-b83a-11e9-9134-698dccb744fb.png">
